doc/ipython.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH IPYTHON 1 "November 30, 2004"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7) and groff_man(7)
.\" .SH        section heading
.\" .SS        secondary section heading
.\"
.\"
.\" To preview this page as plain text: nroff -man ipython.1
.\"
.SH NAME
ipython \- An Enhanced Interactive Python
.SH SYNOPSIS
.B ipython
.RI [ options ] " files" ...
.SH DESCRIPTION
An interactive Python shell with automatic history (input and output),
dynamic object introspection, easier configuration, command
completion, access to the system shell, integration with numerical and
scientific computing tools, and more.
.SH SPECIAL THREADING OPTIONS
The following special options are ONLY valid at the beginning of the command
line, and not later.  This is because they control the initialization of
ipython itself, before the normal option-handling mechanism is active.
.TP
.B \-gthread, \-qthread, \-q4thread, \-wthread, \-pylab
Only ONE of these can be given, and it can only be given as the first option
passed to IPython (it will have no effect in any other position).  They provide
threading support for the GTK, QT3, QT4 and WXWidgets toolkits, for the
matplotlib library and Twisted reactor.
.br
.sp 1
With any of the first four options, IPython starts running a separate thread
for the graphical toolkit's operation, so that you can open and control
graphical elements from within an IPython command line, without blocking.  All
four provide essentially the same functionality, respectively for GTK, QT3, QT4
and WXWidgets (via their Python interfaces).
.br
.sp 1
Note that with \-wthread, you can additionally use the \-wxversion option to
request a specific version of wx to be used.  This requires that you have the
wxversion Python module installed, which is part of recent wxPython
distributions.
.br
.sp 1
If \-pylab is given, IPython loads special support for the matplotlib library
(http://matplotlib.sourceforge.net), allowing interactive usage of any of its
backends as defined in the user's .matplotlibrc file.  It automatically
activates GTK, QT or WX threading for IPyhton if the choice of matplotlib
backend requires it.  It also modifies the %run command to correctly execute
(without blocking) any matplotlib-based script which calls show() at the end.
.TP
.B \-tk
The \-g/q/q4/wthread options, and \-pylab (if matplotlib is configured to use
GTK, QT or WX), will normally block Tk graphical interfaces.  This means that
when GTK, QT or WX threading is active, any attempt to open a Tk GUI will
result in a dead window, and possibly cause the Python interpreter to crash.
An extra option, \-tk, is available to address this issue.  It can ONLY be
given as a SECOND option after any of the above (\-gthread, \-qthread,
\-wthread or \-pylab).
.br
.sp 1
If \-tk is given, IPython will try to coordinate Tk threading with GTK, QT or
WX.  This is however potentially unreliable, and you will have to test on your
platform and Python configuration to determine whether it works for you.
Debian users have reported success, apparently due to the fact that Debian
builds all of Tcl, Tk, Tkinter and Python with pthreads support.  Under other
Linux environments (such as Fedora Core 2), this option has caused random
crashes and lockups of the Python interpreter.  Under other operating systems
(Mac OSX and Windows), you'll need to try it to find out, since currently no
user reports are available.
.br
.sp 1
There is unfortunately no way for IPython to determine at runtime whether \-tk
will work reliably or not, so you will need to do some experiments before
relying on it for regular work.
.
.SH REGULAR OPTIONS
After the above threading options have been given, regular options can follow
in any order.  All options can be abbreviated to their shortest non-ambiguous
form and are case-sensitive.  One or two dashes can be used.  Some options
have an alternate short form, indicated after a |.
.br
.sp 1
Most options can also be set from your ipythonrc configuration file.
See the provided examples for assistance.  Options given on the
commandline override the values set in the ipythonrc file.
.br
.sp 1
All options with a [no] prepended can be specified in negated form
(\-nooption instead of \-option) to turn the feature off.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-autocall <val>
Make IPython automatically call any callable object even if you didn't type
explicit parentheses. For example, 'str 43' becomes
str(43) automatically.  The value can be '0' to disable the
feature, '1' for 'smart' autocall, where it is not applied if
there are no more arguments on the line, and '2' for 'full'
autocall, where all callable objects are automatically called
(even if no arguments are present).  The default is '1'.
.TP
.B \-[no]autoindent
Turn automatic indentation on/off.
.TP
.B \-[no]automagic
Make magic commands automatic (without needing their first character
to be %).  Type %magic at the IPython prompt for more information.
.TP
.B \-[no]autoedit_syntax
When a syntax error occurs after editing a file, automatically open the file
to the trouble causing line for convenient fixing.
.TP
.B \-[no]banner
Print the intial information banner (default on).
.TP
.B \-c <command>
Execute the given command string, and set sys.argv to ['c'].  This is similar
to the \-c option in the normal Python interpreter.
.TP
.B \-cache_size|cs <n>
Size of the output cache (maximum number of entries to hold in
memory).  The default is 1000, you can change it permanently in your
config file.  Setting it to 0 completely disables the caching system,
and the minimum value accepted is 20 (if you provide a value less than
20, it is reset to 0 and a warning is issued).  This limit is defined
because otherwise you'll spend more time re-flushing a too small cache
than working.
.TP
.B \-classic|cl
Gives IPython a similar feel to the classic Python prompt.
.TP
.B \-colors <scheme>
Color scheme for prompts and exception reporting.  Currently
implemented: NoColor, Linux, and LightBG.
.TP
.B \-[no]color_info
IPython can display information about objects via a set of functions,
and optionally can use colors for this, syntax highlighting source
code and various other elements.  However, because this information is
passed through a pager (like 'less') and many pagers get confused with
color codes, this option is off by default.  You can test it and turn
it on permanently in your ipythonrc file if it works for you.  As a
reference, the 'less' pager supplied with Mandrake 8.2 works ok, but
that in RedHat 7.2 doesn't.
.br
.sp 1
Test it and turn it on permanently if it works with your system.  The
magic function @color_info allows you to toggle this interactively for
testing.
.TP
.B \-[no]confirm_exit
Set to confirm when you try to exit IPython with an EOF (Control-D in
Unix, Control-Z/Enter in Windows). Note that using the magic functions
@Exit or @Quit you can force a direct exit, bypassing any
confirmation.
.TP
.B \-[no]debug
Show information about the loading process. Very useful to pin down
problems with your configuration files or to get details about session
restores.
.TP
.B \-[no]deep_reload
IPython can use the deep_reload module which reloads changes in
modules recursively (it replaces the reload() function, so you don't
need to change anything to use it). deep_reload() forces a full reload
of modules whose code may have changed, which the default reload()
function does not.
.br
.sp 1
When deep_reload is off, IPython will use the normal reload(), but
deep_reload will still be available as dreload(). This feature is off
by default [which means that you have both normal reload() and
dreload()].
.TP
.B \-editor <name>
Which editor to use with the @edit command. By default, IPython will
honor your EDITOR environment variable (if not set, vi is the Unix
default and notepad the Windows one). Since this editor is invoked on
the fly by IPython and is meant for editing small code snippets, you
may want to use a small, lightweight editor here (in case your default
EDITOR is something like Emacs).
.TP
.B \-ipythondir <name>
The name of your IPython configuration directory IPYTHONDIR.  This can
also be specified through the environment variable IPYTHONDIR.
.TP
.B \-log|l
Generate a log file of all input. The file is named ipython_log.py in your
current directory (which prevents logs from multiple IPython sessions from
trampling each other). You can use this to later restore a session by loading
your logfile as a file to be executed with option -logplay (see below).
.TP
.B \-logfile|lf
Specify the name of your logfile.
.TP
.B \-logplay|lp
Replay a previous log. For restoring a session as close as possible to
the state you left it in, use this option (don't just run the
logfile). With \-logplay, IPython will try to reconstruct the previous
working environment in full, not just execute the commands in the
logfile.
.br
.sp 1
When a session is restored, logging is automatically turned on again
with the name of the logfile it was invoked with (it is read from the
log header). So once you've turned logging on for a session, you can
quit IPython and reload it as many times as you want and it will
continue to log its history and restore from the beginning every time.
.br
.sp 1
Caveats: there are limitations in this option. The history variables
_i*,_* and _dh don't get restored properly. In the future we will try
to implement full session saving by writing and retrieving a
snapshot of the memory state of IPython. But our first attempts
failed because of inherent limitations of Python's Pickle module, so
this may have to wait.
.TP
.B \-[no]messages
Print messages which IPython collects about its startup process
(default on).
.TP
.B \-[no]pdb
Automatically call the pdb debugger after every uncaught exception. If
you are used to debugging using pdb, this puts you automatically
inside of it after any call (either in IPython or in code called by
it) which triggers an exception which goes uncaught.
.TP
.B \-pydb
Makes IPython use the third party "pydb" package as debugger,
instead of pdb. Requires that pydb is installed.
.TP
.B \-[no]pprint
IPython can optionally use the pprint (pretty printer) module for
displaying results. pprint tends to give a nicer display of nested
data structures. If you like it, you can turn it on permanently in
your config file (default off).
.TP
.B \-profile|p <name>
Assume that your config file is ipythonrc-<name> (looks in current dir
first, then in IPYTHONDIR). This is a quick way to keep and load
multiple config files for different tasks, especially if you use the
include option of config files. You can keep a basic
IPYTHONDIR/ipythonrc file and then have other 'profiles' which include
this one and load extra things for particular tasks. For example:
.br
.sp 1
1) $HOME/.ipython/ipythonrc : load basic things you always want.
.br
2) $HOME/.ipython/ipythonrc-math : load (1) and basic math-related
modules.
.br
3) $HOME/.ipython/ipythonrc-numeric : load (1) and Numeric and
plotting modules.
.br
.sp 1
Since it is possible to create an endless loop by having circular file
inclusions, IPython will stop if it reaches 15 recursive inclusions.
.TP
.B \-prompt_in1|pi1 <string>
Specify the string used for input prompts. Note that if you are using
numbered prompts, the number is represented with a '\\#' in the
string. Don't forget to quote strings with spaces embedded in
them. Default: 'In [\\#]: '.
.br
.sp 1
Most bash-like escapes can be used to customize IPython's prompts, as well as
a few additional ones which are IPython-specific.  All valid prompt escapes
are described in detail in the Customization section of the IPython HTML/PDF
manual.
.TP
.B \-prompt_in2|pi2 <string>
Similar to the previous option, but used for the continuation prompts. The
special sequence '\\D' is similar to '\\#', but with all digits replaced dots
(so you can have your continuation prompt aligned with your input
prompt). Default: '   .\\D.: ' (note three spaces at the start for alignment
with 'In [\\#]').
.TP
.B \-prompt_out|po <string>
String used for output prompts, also uses numbers like prompt_in1.
Default: 'Out[\\#]:'.
.TP
.B \-quick
Start in bare bones mode (no config file loaded).
.TP
.B \-rcfile <name>
Name of your IPython resource configuration file.  normally IPython
loads ipythonrc (from current directory) or IPYTHONDIR/ipythonrc.  If
the loading of your config file fails, IPython starts with a bare
bones configuration (no modules loaded at all).
.TP
.B \-[no]readline
Use the readline library, which is needed to support name completion
and command history, among other things. It is enabled by default, but
may cause problems for users of X/Emacs in Python comint or shell
buffers.
.br
.sp 1
Note that emacs 'eterm' buffers (opened with M-x term) support
IPython's readline and syntax coloring fine, only 'emacs' (M-x shell
and C-c !)  buffers do not.
.TP
.B \-screen_length|sl <n>
Number of lines of your screen.  This is used to control printing of
very long strings.  Strings longer than this number of lines will be
sent through a pager instead of directly printed.
.br
.sp 1
The default value for this is 0, which means IPython will auto-detect
your screen size every time it needs to print certain potentially long
strings (this doesn't change the behavior of the 'print' keyword, it's
only triggered internally). If for some reason this isn't working well
(it needs curses support), specify it yourself. Otherwise don't change
the default.
.TP
.B \-separate_in|si <string>
Separator before input prompts.  Default '\n'.
.TP
.B \-separate_out|so <string>
Separator before output prompts.  Default: 0 (nothing).
.TP
.B \-separate_out2|so2 <string>
Separator after output prompts.  Default: 0 (nothing).
.TP
.B \-nosep
Shorthand for '\-separate_in 0 \-separate_out 0 \-separate_out2 0'.
Simply removes all input/output separators.
.TP
.B \-upgrade
Allows you to upgrade your IPYTHONDIR configuration when you install a
new version of IPython.  Since new versions may include new command
lines options or example files, this copies updated ipythonrc-type
files.  However, it backs up (with a .old extension) all files which
it overwrites so that you can merge back any custimizations you might
have in your personal files.
.TP
.B \-Version
Print version information and exit.
.TP
.B -wxversion <string>
Select a specific version of wxPython (used in conjunction with
\-wthread). Requires the wxversion module, part of recent wxPython
distributions.
.TP
.B \-xmode <modename>
Mode for exception reporting.  The valid modes are Plain, Context, and
Verbose.
.br
.sp 1
\- Plain: similar to python's normal traceback printing.
.br
.sp 1

ChangeLog
This file used to be symlink in SVN, but now it serves as a *warning* so that
nobody creates symlinks to the bzr repository



docs/emacs/ipython.el
;;; ipython.el --- Adds support for IPython to python-mode.el

;; Copyright (C) 2002, 2003, 2004, 2005 Alexander Schmolck
;; Author:        Alexander Schmolck
;; Keywords:      ipython python languages oop
;; URL:           http://ipython.scipy.org
;; Compatibility: Emacs21, XEmacs21
;; FIXME: #$@! INPUT RING
(defconst ipython-version "$Revision: 2927 $"
  "VC version number.")

;;; Commentary 
;; This library makes all the functionality python-mode has when running with
;; the normal python-interpreter available for ipython, too. It also enables a
;; persistent py-shell command history across sessions (if you exit python
;; with C-d in py-shell) and defines the command `ipython-to-doctest', which
;; can be used to convert bits of a ipython session into something that can be
;; used for doctests. To install, put this file somewhere in your emacs
;; `load-path' [1] and add the following line to your ~/.emacs file (the first
;; line only needed if the default (``"ipython"``) is wrong)::
;;
;;   (setq ipython-command "/SOME-PATH/ipython")
;;   (require 'ipython)
;;
;; Ipython will be set as the default python shell, but only if the ipython
;; executable is in the path. For ipython sessions autocompletion with <tab>
;; is also enabled (experimental feature!). Please also note that all the
;; terminal functions in py-shell are handled by emacs's comint, **not** by
;; (i)python, so importing readline etc. will have 0 effect.
;;
;; To start an interactive ipython session run `py-shell' with ``M-x py-shell``
;; (or the default keybinding ``C-c C-!``).
;;
;; NOTE: This mode is currently somewhat alpha and although I hope that it
;; will work fine for most cases, doing certain things (like the
;; autocompletion and a decent scheme to switch between python interpreters)
;; properly will also require changes to ipython that will likely have to wait
;; for a larger rewrite scheduled some time in the future.
;; 
;; Also note that you currently NEED THE CVS VERSION OF PYTHON.EL.
;;
;; Further note that I don't know whether this runs under windows or not and
;; that if it doesn't I can't really help much, not being afflicted myself.
;;
;;
;; Hints for effective usage
;; -------------------------
;;
;; - IMO the best feature by far of the ipython/emacs combo is how much easier it
;;   makes it to find and fix bugs thanks to the ``%pdb on``/ pdbtrack combo. Try
;;   it: first in the ipython to shell do ``%pdb on`` then do something that will
;;   raise an exception (FIXME nice example) -- and be amazed how easy it is to
;;   inspect the live objects in each stack frames and to jump to the
;;   corresponding sourcecode locations as you walk up and down the stack trace
;;   (even without ``%pdb on`` you can always use ``C-c -`` (`py-up-exception')
;;   to jump to the corresponding source code locations).
;;
;; - emacs gives you much more powerful commandline editing and output searching
;;   capabilities than ipython-standalone -- isearch is your friend if you
;;   quickly want to print 'DEBUG ...' to stdout out etc.
;;
;; - This is not really specific to ipython, but for more convenient history
;;   access you might want to add something like the following to *the beggining*
;;   of your ``.emacs`` (if you want behavior that's more similar to stand-alone
;;   ipython, you can change ``meta p`` etc. for ``control p``)::
;;
;;         (require 'comint)
;;         (define-key comint-mode-map [(meta p)] 
;;           'comint-previous-matching-input-from-input)
;;         (define-key comint-mode-map [(meta n)] 
;;           'comint-next-matching-input-from-input)
;;         (define-key comint-mode-map [(control meta n)]
;;            'comint-next-input)
;;         (define-key comint-mode-map [(control meta p)]
;;            'comint-previous-input)
;;
;; - Be aware that if you customize py-python-command previously, this value
;;   will override what ipython.el does (because loading the customization
;;   variables comes later).
;;
;; Please send comments and feedback to the ipython-list
;; (<ipython-user@scipy.net>) where I (a.s.) or someone else will try to
;; answer them (it helps if you specify your emacs version, OS etc; 
;; familiarity with <http://www.catb.org/~esr/faqs/smart-questions.html> might
;; speed up things further).
;;
;; Footnotes:
;;
;;     [1] If you don't know what `load-path' is, C-h v load-path will tell
;;     you; if required you can also add a new directory. So assuming that
;;     ipython.el resides in ~/el/, put this in your emacs:
;;
;;
;;           (add-to-list 'load-path "~/el")
;;           (setq ipython-command "/some-path/ipython")
;;           (require 'ipython)
;;
;;
;;
;;
;; TODO:
;;      - do autocompletion properly
;;      - implement a proper switching between python interpreters
;;
;; BUGS:
;;      - neither::
;;
;;         (py-shell "-c print 'FOOBAR'")
;;       
;;        nor::
;;       
;;         (let ((py-python-command-args (append py-python-command-args 
;;                                              '("-c" "print 'FOOBAR'"))))
;;           (py-shell))
;;
;;        seem to print anything as they should
;;
;;      - look into init priority issues with `py-python-command' (if it's set
;;        via custom)


;;; Code
(require 'cl)
(require 'shell)
(require 'executable)
(require 'ansi-color)

(defcustom ipython-command "ipython"
  "*Shell command used to start ipython."
  :type 'string 
  :group 'python)

;; Users can set this to nil
(defvar py-shell-initial-switch-buffers t
  "If nil, don't switch to the *Python* buffer on the first call to
  `py-shell'.")

(defvar ipython-backup-of-py-python-command nil
  "HACK")

  
(defvar ipython-de-input-prompt-regexp "\\(?:
In \\[[0-9]+\\]: *.*
----+> \\(.*
\\)[\n]?\\)\\|\\(?:
In \\[[0-9]+\\]: *\\(.*
\\)\\)\\|^[ ]\\{3\\}[.]\\{3,\\}: *\\(.*
\\)"
  "A regular expression to match the IPython input prompt and the python
command after it. The first match group is for a command that is rewritten,
the second for a 'normal' command, and the third for a multiline command.")
(defvar ipython-de-output-prompt-regexp "^Out\\[[0-9]+\\]: "
  "A regular expression to match the output prompt of IPython.")


(if (not (executable-find ipython-command))
    (message (format "Can't find executable %s - ipython.el *NOT* activated!!!"
                     ipython-command))
    ;; XXX load python-mode, so that we can screw around with its variables
    ;; this has the disadvantage that python-mode is loaded even if no
    ;; python-file is ever edited etc. but it means that `py-shell' works
    ;; without loading a python-file first. Obviously screwing around with
    ;; python-mode's variables like this is a mess, but well.
    (require 'python-mode)
    ;; turn on ansi colors for ipython and activate completion
    (defun ipython-shell-hook ()
      ;; the following is to synchronize dir-changes
      (make-local-variable 'shell-dirstack)
      (setq shell-dirstack nil)
      (make-local-variable 'shell-last-dir)
      (setq shell-last-dir nil)
      (make-local-variable 'shell-dirtrackp)
      (setq shell-dirtrackp t)
      (add-hook 'comint-input-filter-functions 'shell-directory-tracker nil t)

      (ansi-color-for-comint-mode-on)
      (define-key py-shell-map [tab] 'ipython-complete)
      ;; Add this so that tab-completion works both in X11 frames and inside
      ;; terminals (such as when emacs is called with -nw).
      (define-key py-shell-map "\t" 'ipython-complete)
      ;;XXX this is really just a cheap hack, it only completes symbols in the
      ;;interactive session -- useful nonetheless.
      (define-key py-mode-map [(meta tab)] 'ipython-complete)
      
      )
    (add-hook 'py-shell-hook 'ipython-shell-hook)
    ;; Regular expression that describes tracebacks for IPython in context and
    ;; verbose mode. 
  
    ;;Adapt python-mode settings for ipython.
    ;; (this works for %xmode 'verbose' or 'context')

    ;; XXX putative regexps for syntax errors; unfortunately the 
    ;;     current python-mode traceback-line-re scheme is too primitive,
    ;;     so it's either matching syntax errors, *or* everything else
    ;;     (XXX: should ask Fernando for a change)
    ;;"^   File \"\\(.*?\\)\", line \\([0-9]+\\).*\n.*\n.*\nSyntaxError:"
    ;;^   File \"\\(.*?\\)\", line \\([0-9]+\\)"

    (setq py-traceback-line-re
          "\\(^[^\t >].+?\\.py\\).*\n   +[0-9]+[^\00]*?\n-+> \\([0-9]+\\)+")

    
    ;; Recognize the ipython pdb, whose prompt is 'ipdb>' or  'ipydb>'
    ;;instead of '(Pdb)'
    (setq py-pdbtrack-input-prompt "\n[(<]*[Ii]?[Pp]y?db[>)]+ ")
    (setq pydb-pydbtrack-input-prompt "\n[(]*ipydb[>)]+ ")
    
    (setq py-shell-input-prompt-1-regexp "^In \\[[0-9]+\\]: *"
          py-shell-input-prompt-2-regexp "^   [.][.][.]+: *" )
    ;; select a suitable color-scheme
    (unless (member "-colors" py-python-command-args)
      (setq py-python-command-args 
            (nconc py-python-command-args 
                   (list "-colors"
                         (cond  
                           ((eq frame-background-mode 'dark)
                            "Linux")
                           ((eq frame-background-mode 'light)
                            "LightBG")
                           (t ; default (backg-mode isn't always set by XEmacs)
                            "LightBG"))))))
    (unless (equal ipython-backup-of-py-python-command py-python-command)
      (setq ipython-backup-of-py-python-command py-python-command))
    (setq py-python-command ipython-command))


;; MODIFY py-shell so that it loads the editing history
(defadvice py-shell (around py-shell-with-history)
  "Add persistent command-history support (in
$PYTHONHISTORY (or \"~/.ipython/history\", if we use IPython)). Also, if
`py-shell-initial-switch-buffers' is nil, it only switches to *Python* if that
buffer already exists."
  (if (comint-check-proc "*Python*")
      ad-do-it
    (setq comint-input-ring-file-name
          (if (string-equal py-python-command ipython-command)
              (concat (or (getenv "IPYTHONDIR") "~/.ipython") "/history")
            (or (getenv "PYTHONHISTORY") "~/.python-history.py")))
    (comint-read-input-ring t)
    (let ((buf (current-buffer)))
      ad-do-it
      (unless py-shell-initial-switch-buffers
        (switch-to-buffer-other-window buf)))))
(ad-activate 'py-shell)
;; (defadvice py-execute-region (before py-execute-buffer-ensure-process)
;;   "HACK: test that ipython is already running before executing something.
;;   Doing this properly seems not worth the bother (unless people actually
;;   request it)."
;; (unless (comint-check-proc "*Python*")
;;     (error "Sorry you have to first do M-x py-shell to send something to ipython.")))
;; (ad-activate 'py-execute-region)

(defadvice py-execute-region (around py-execute-buffer-ensure-process)
  "HACK: if `py-shell' is not active or ASYNC is explicitly desired, fall back
  to python instead of ipython." 
  (let ((py-which-shell (if (and (comint-check-proc "*Python*") (not async))
			    py-python-command
			  ipython-backup-of-py-python-command)))
    ad-do-it))
(ad-activate 'py-execute-region)

(defun ipython-to-doctest (start end)
  "Transform a cut-and-pasted bit from an IPython session into something that
looks like it came from a normal interactive python session, so that it can
be used in doctests. Example:


    In [1]: import sys
    
    In [2]: sys.stdout.write 'Hi!\n'
    ------> sys.stdout.write ('Hi!\n')
    Hi!
    
    In [3]: 3 + 4
    Out[3]: 7
    
gets converted to:

    >>> import sys
    >>> sys.stdout.write ('Hi!\n')
    Hi!
    >>> 3 + 4
    7

"
  (interactive "*r\n")
  ;(message (format "###DEBUG s:%de:%d" start end))
  (save-excursion
    (save-match-data
      ;; replace ``In [3]: bla`` with ``>>> bla`` and 
      ;;         ``... :   bla`` with ``...    bla``
      (goto-char start)
      (while (re-search-forward ipython-de-input-prompt-regexp end t)
        ;(message "finding 1")
        (cond ((match-string 3)         ;continued
               (replace-match "... \\3" t nil))
              (t
               (replace-match ">>> \\1\\2" t nil))))
      ;; replace ``
      (goto-char start)
      (while (re-search-forward ipython-de-output-prompt-regexp end t)
        (replace-match "" t nil)))))

(defvar ipython-completion-command-string 
  "print ';'.join(__IP.Completer.all_completions('%s')) #PYTHON-MODE SILENT\n"
  "The string send to ipython to query for all possible completions")


;; xemacs doesn't have `comint-preoutput-filter-functions' so we'll try the
;; following wonderful hack to work around this case
(if (featurep 'xemacs)
    ;;xemacs
    (defun ipython-complete ()
      "Try to complete the python symbol before point. Only knows about the stuff
in the current *Python* session."
      (interactive)
      (let* ((ugly-return nil)
             (sep ";")
             (python-process (or (get-buffer-process (current-buffer))
                                 ;XXX hack for .py buffers
                                 (get-process py-which-bufname)))
             ;; XXX currently we go backwards to find the beginning of an
             ;; expression part; a more powerful approach in the future might be
             ;; to let ipython have the complete line, so that context can be used
             ;; to do things like filename completion etc.
             (beg (save-excursion (skip-chars-backward "a-z0-9A-Z_." (point-at-bol))
                                  (point)))
             (end (point))
             (pattern (buffer-substring-no-properties beg end))
             (completions nil)
             (completion-table nil)
             completion
             (comint-output-filter-functions
              (append comint-output-filter-functions 
                      '(ansi-color-filter-apply
                        (lambda (string) 
                                        ;(message (format "DEBUG filtering: %s" string))
                          (setq ugly-return (concat ugly-return string))
                          (delete-region comint-last-output-start 
                                         (process-mark (get-buffer-process (current-buffer)))))))))
                                        ;(message (format "#DEBUG pattern: '%s'" pattern))
        (process-send-string python-process 
                              (format ipython-completion-command-string pattern))
        (accept-process-output python-process)
                                        ;(message (format "DEBUG return: %s" ugly-return))
        (setq completions 
              (split-string (substring ugly-return 0 (position ?\n ugly-return)) sep))
        (setq completion-table (loop for str in completions
                                     collect (list str nil)))
        (setq completion (try-completion pattern completion-table))
        (cond ((eq completion t))
              ((null completion)
               (message "Can't find completion for \"%s\"" pattern)
               (ding))
              ((not (string= pattern completion))
               (delete-region beg end)
               (insert completion))
              (t
               (message "Making completion list...")
               (with-output-to-temp-buffer "*Python Completions*"
                 (display-completion-list (all-completions pattern completion-table)))
               (message "Making completion list...%s" "done")))))
  ;; emacs
  (defun ipython-complete ()
    "Try to complete the python symbol before point. Only knows about the stuff
in the current *Python* session."
    (interactive)
    (let* ((ugly-return nil)
           (sep ";")
           (python-process (or (get-buffer-process (current-buffer))
                                        ;XXX hack for .py buffers
                               (get-process py-which-bufname)))
           ;; XXX currently we go backwards to find the beginning of an
           ;; expression part; a more powerful approach in the future might be
           ;; to let ipython have the complete line, so that context can be used
           ;; to do things like filename completion etc.
           (beg (save-excursion (skip-chars-backward "a-z0-9A-Z_./" (point-at-bol))
                                (point))) 
           (end (point))
           (pattern (buffer-substring-no-properties beg end))
           (completions nil)
           (completion-table nil)
           completion
         (comint-preoutput-filter-functions
          (append comint-preoutput-filter-functions 
                  '(ansi-color-filter-apply
                    (lambda (string) 
                      (setq ugly-return (concat ugly-return string))
                      "")))))
      (process-send-string python-process 
                            (format ipython-completion-command-string pattern))
      (accept-process-output python-process)
      (setq completions 
            (split-string (substring ugly-return 0 (position ?\n ugly-return)) sep))
                                        ;(message (format "DEBUG completions: %S" completions))
      (setq completion-table (loop for str in completions
                                   collect (list str nil)))
      (setq completion (try-completion pattern completion-table))
      (cond ((eq completion t))
            ((null completion)
             (message "Can't find completion for \"%s\"" pattern)
             (ding))
            ((not (string= pattern completion))
             (delete-region beg end)
             (insert completion))
            (t
             (message "Making completion list...")
             (with-output-to-temp-buffer "*IPython Completions*"
               (display-completion-list (all-completions pattern completion-table)))
             (message "Making completion list...%s" "done")))))
)

;;; autoindent support: patch sent in by Jin Liu <m.liu.jin@gmail.com>,
;;; originally written by doxgen@newsmth.net
;;; Minor modifications by fperez for xemacs compatibility.

(defvar ipython-autoindent t
 "If non-nil, enable autoindent for IPython shell through python-mode.")

(defvar ipython-indenting-buffer-name "*IPython Indentation Calculation*"
 "Temporary buffer for indenting multiline statement.")

(defun ipython-get-indenting-buffer ()
 "Return a temporary buffer set in python-mode. Create one if necessary."
 (let ((buf (get-buffer-create ipython-indenting-buffer-name)))
   (set-buffer buf)
   (unless (eq major-mode 'python-mode)
     (python-mode))
   buf))

(defvar ipython-indentation-string nil
 "Indentation for the next line in a multiline statement.")

(defun ipython-send-and-indent ()
 "Send the current line to IPython, and calculate the indentation for
the next line."
 (interactive)
 (if ipython-autoindent
     (let ((line (buffer-substring (point-at-bol) (point)))
           (after-prompt1)
           (after-prompt2))
       (save-excursion
           (comint-bol t)
           (if (looking-at py-shell-input-prompt-1-regexp)
               (setq after-prompt1 t)
             (setq after-prompt2 (looking-at py-shell-input-prompt-2-regexp)))
           (with-current-buffer (ipython-get-indenting-buffer)
             (when after-prompt1
               (erase-buffer))
             (when (or after-prompt1 after-prompt2)
               (delete-region (point-at-bol) (point))
               (insert line)
               (newline-and-indent))))))
 ;; send input line to ipython interpreter
 (comint-send-input))

(defun ipython-indentation-hook (string)
 "Insert indentation string if py-shell-input-prompt-2-regexp
matches last process output."
 (let* ((start-marker (or comint-last-output-start
                          (point-min-marker)))
        (end-marker (process-mark (get-buffer-process (current-buffer))))
        (text (ansi-color-filter-apply (buffer-substring start-marker end-marker))))
   ;; XXX if `text' matches both pattern, it MUST be the last prompt-2
   (when (and (string-match py-shell-input-prompt-2-regexp text)
	      (not (string-match "\n$" text)))
     (with-current-buffer (ipython-get-indenting-buffer)
       (setq ipython-indentation-string
	     (buffer-substring (point-at-bol) (point))))
     (goto-char end-marker)
     (insert ipython-indentation-string)
     (setq ipython-indentation-string nil))))

(add-hook 'py-shell-hook
         (lambda ()
           (add-hook 'comint-output-filter-functions
                     'ipython-indentation-hook)))

(define-key py-shell-map (kbd "RET") 'ipython-send-and-indent)
;;; / end autoindent support

(provide 'ipython)
